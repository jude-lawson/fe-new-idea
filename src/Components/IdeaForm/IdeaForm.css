.idea-form {
  background-image: linear-gradient(
    to top,
    #4381c1,
    #289dc5,
    #50b4be,
    #85c7b7,
    #b8d8ba
  );
  margin: 10rem auto 4rem;
  border-radius: 20px;
  padding: 2rem 4rem;
  display: -ms-flexbox;
  display: flex;
  flex-direction: column;
  width: 65rem;
  height: 61rem;
}

.idea--form-title {
  text-align: center;
  margin-bottom: 2rem;
  color: var(--primary-blue);
}

.idea--title-input {
  margin-bottom: 2rem;
  height: 3rem;
  font-size: 16px;
  padding: 3px;
}

.idea--body-input {
  height: 49rem;
  font-size: 14px;
  line-height: 1.4;
  padding: 3px 5px;
}

.idea--form-btn {
  margin-top: 4rem;
  color: var(--primary-blue);
  background: var(--primary-white);
  width: 12rem;
  padding: 1rem;
  margin: 2rem auto;
  font-weight: 700;
  border-radius: 23px;
}

.idea--form-btn:hover {
  background: #a7d3b9;
}

@media screen and (max-width: 825px) {
  .idea-form {
    width: 80%;
  }
}

@media screen and (max-width: 500px) {
  .idea-form {
    width: 95%;
    margin: 0rem auto;
    border-radius: 0px;
    height: calc(100vh - 8rem);
    padding: 2.5%;
    display: -ms-flexbox;
    display: flex;
    -ms-flex-direction: column;
    flex-direction: column;
  }

  .idea--body-input {
    height: calc(66vh - 4rem);
    font-size: 14px;
    line-height: 1.4;
    padding: 3px 5px;
  }
}