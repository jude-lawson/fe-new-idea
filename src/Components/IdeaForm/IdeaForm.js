import React, {Component} from 'react';

export class IdeaForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      body: ''
    };
  }

  handleChange = (event) => {
    const {name, value} = event.target;
    this.setState({
      [name]: value
    });
  }

  handleSubmit = (event) => {
    event.preventDefault();
  }

  render() {
    return (
      <form>
        <input 
          type="text"
          className="article--title-input"
          placeholder="title"
          onChange={this.handleChange}
          name="title"
        />
        <textarea
          className="article--body-input"
          placeholder="Description"
          name="body"
          onChange={this.handleChange}
        />
        <button>SUBMIT</button>
      </form>
    );
  }
}