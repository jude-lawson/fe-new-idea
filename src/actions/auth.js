export const userLoginAction = user => ({
  type: 'LOGIN',
  user
});