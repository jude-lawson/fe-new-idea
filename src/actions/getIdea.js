export const getIdeaAction = idea => ({
  type: 'GET_IDEA',
  idea
});