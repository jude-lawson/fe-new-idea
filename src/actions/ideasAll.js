export const getIdeas = ideas => ({
  type: 'GET_IDEAS',
  ideas
});