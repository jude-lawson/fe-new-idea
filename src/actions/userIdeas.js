export const retrieveMyIdeas = ideas => ({
  type: 'RETRIEVE_MY_IDEAS',
  ideas
});