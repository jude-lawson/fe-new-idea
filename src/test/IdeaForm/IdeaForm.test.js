import { IdeaForm, mapStateToProps, mapDispatchToProps } from '../../Components/IdeaForm/IdeaForm';
import { shallow, mount } from 'enzyme';
import React from 'react';
import { addIdea } from '../../actions/idea';
import { storageMock } from '../test-helpers/localstroage';

describe('IdeaForm', () => {
  let spy;
  let wrapper;
  const mockFunc = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<IdeaForm addIdea={mockFunc}/>);
    spy = jest.spyOn(wrapper.instance(), 'handleChange');
    window.localStorage = storageMock()

  });

  it('should match the snapshot', () => {
    let wrapper = shallow(<IdeaForm />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it('should update state of title', () => {
      const mockEvent = {
        target: {
          value: 'h',
          name: 'title'
        }
      };

      wrapper.find('.idea--title-input').simulate('change', {target: {value: 'h', name: 'title'}});
      wrapper.instance().handleChange(mockEvent);

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state('title')).toEqual('h');
    });

    it('should update state of body', () => {
      const mockEvent = {
        target: {
          value: 'tech things',
          name: 'body'
        }
      };
      wrapper.instance().handleChange(mockEvent);
      wrapper.find('input').last().simulate('change', {target: {value: 'tech things', name: 'body'}});

      expect(spy).toHaveBeenCalled();
      expect(wrapper.state('body')).toEqual('tech things');
    });
  });

  describe('handleSubmit', () => {
    it('should call handleSubmit on click', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleSubmit');

      const mockEvent = {
        preventDefault: jest.fn()
      }
      wrapper.setState({
        title: 'hey',
        body: 'you'
      });
      window.fetch = jest.fn().mockImplementation(() => 
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve('Idea created successfully')
      }))

      wrapper.find('form').simulate('submit', mockEvent)
      expect(spy).toHaveBeenCalled();
    });
  })

  describe('mapStateToProps', () => {
    it('should create a props object with the correct keys', () => {
      const mockState = {
        ideas: [
          {
            title: 'Latest and Greatest',
            body: 'Is it or is it not the latest and greatest?',
            id: 11
          }
        ]
      }
      const expected = {
        ideas: [
        {
          title: 'Latest and Greatest',
          body: 'Is it or is it not the latest and greatest?',
          id: 11
        }
        ]
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);

    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when using addIdea from MDTP', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = addIdea({title: 'Tech', body: 'Tech body', id: 100});
      const mappedToProps = mapDispatchToProps(mockDispatch);
      mappedToProps.addIdea({title: 'Tech', body: 'Tech body', id: 100});

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});