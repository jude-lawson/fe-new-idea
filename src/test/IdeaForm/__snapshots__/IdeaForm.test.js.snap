// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeaForm should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <IdeaForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Header) />,
        <form
          className="idea-form"
          onSubmit={[Function]}
>
          <h1
                    className="idea--form-title"
          >
                    Create A New Idea
          </h1>
          <input
                    className="idea--title-input"
                    name="title"
                    onChange={[Function]}
                    placeholder="Title"
                    type="text"
                    value=""
          />
          <textarea
                    className="idea--body-input"
                    name="body"
                    onChange={[Function]}
                    placeholder="Description"
                    value=""
          />
          <button
                    className="idea--form-btn"
          >
                    SUBMIT
          </button>
</form>,
      ],
      "className": "app-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="idea--form-title"
>
              Create A New Idea
</h1>,
            <input
              className="idea--title-input"
              name="title"
              onChange={[Function]}
              placeholder="Title"
              type="text"
              value=""
/>,
            <textarea
              className="idea--body-input"
              name="body"
              onChange={[Function]}
              placeholder="Description"
              value=""
/>,
            <button
              className="idea--form-btn"
>
              SUBMIT
</button>,
          ],
          "className": "idea-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Create A New Idea",
              "className": "idea--form-title",
            },
            "ref": null,
            "rendered": "Create A New Idea",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "idea--title-input",
              "name": "title",
              "onChange": [Function],
              "placeholder": "Title",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "idea--body-input",
              "name": "body",
              "onChange": [Function],
              "placeholder": "Description",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "SUBMIT",
              "className": "idea--form-btn",
            },
            "ref": null,
            "rendered": "SUBMIT",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Header) />,
          <form
            className="idea-form"
            onSubmit={[Function]}
>
            <h1
                        className="idea--form-title"
            >
                        Create A New Idea
            </h1>
            <input
                        className="idea--title-input"
                        name="title"
                        onChange={[Function]}
                        placeholder="Title"
                        type="text"
                        value=""
            />
            <textarea
                        className="idea--body-input"
                        name="body"
                        onChange={[Function]}
                        placeholder="Description"
                        value=""
            />
            <button
                        className="idea--form-btn"
            >
                        SUBMIT
            </button>
</form>,
        ],
        "className": "app-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="idea--form-title"
>
                Create A New Idea
</h1>,
              <input
                className="idea--title-input"
                name="title"
                onChange={[Function]}
                placeholder="Title"
                type="text"
                value=""
/>,
              <textarea
                className="idea--body-input"
                name="body"
                onChange={[Function]}
                placeholder="Description"
                value=""
/>,
              <button
                className="idea--form-btn"
>
                SUBMIT
</button>,
            ],
            "className": "idea-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Create A New Idea",
                "className": "idea--form-title",
              },
              "ref": null,
              "rendered": "Create A New Idea",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "idea--title-input",
                "name": "title",
                "onChange": [Function],
                "placeholder": "Title",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "idea--body-input",
                "name": "body",
                "onChange": [Function],
                "placeholder": "Description",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "SUBMIT",
                "className": "idea--form-btn",
              },
              "ref": null,
              "rendered": "SUBMIT",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
